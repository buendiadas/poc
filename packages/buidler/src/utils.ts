import fs from 'fs-extra';
import path from 'path';
import { BuidlerProvider } from './provider';
import {
  validateConfig,
  resolveConfig,
  defaultConfig,
  solcInputFilename,
  solcOutputFilename,
} from './imports';

export async function addCompilationResult(
  source: string,
  provider: BuidlerProvider
) {
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const loaded = require(source);
  validateConfig(loaded);

  const config = resolveConfig(source, defaultConfig, loaded, []);
  const inputPath = path.join(config.paths.cache, solcInputFilename);
  const outputPath = path.join(config.paths.cache, solcOutputFilename);
  const compilerVersion = config.solc.version;

  const compilerInput = await fs.readJSON(inputPath, {
    encoding: 'utf8',
  });

  const compilerOutput = await fs.readJSON(outputPath, {
    encoding: 'utf8',
  });

  await provider.send('buidler_addCompilationResult', [
    compilerVersion,
    compilerInput,
    compilerOutput,
  ]);
}
