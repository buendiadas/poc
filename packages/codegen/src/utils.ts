export function prettierConfig(cwd: string = process.cwd()) {
  try {
    const prettier = require('prettier');
    return prettier.resolveConfig.sync(cwd);
  } catch (error) {
    return {};
  }
}

export function formatOutput(value: string) {
  try {
    const prettier = require('prettier');
    const defaults = prettierConfig();
    const options = {
      ...defaults,
      parser: 'typescript',
    };

    return prettier.format(value, options);
  } catch (error) {}

  return value;
}
