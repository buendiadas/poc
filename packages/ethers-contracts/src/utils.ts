export function propertyOf<TOr = any>(
  property: string,
  candidates: object[] = [],
): TOr {
  const obj = candidates.find((obj) => obj.hasOwnProperty(property));
  return (obj as any)?.[property] ?? undefined;
}
