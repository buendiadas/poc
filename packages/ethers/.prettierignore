# TODO: Remove this again once prettier supports typescript v4
src/tests/contracts
src/doppelganger.ts
