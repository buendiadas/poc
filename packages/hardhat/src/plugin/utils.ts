import path from 'path';
import fs from 'fs-extra';

export function validateDir(root: string, relative: string) {
  const dir = path.resolve(root, relative);
  if (!dir.startsWith(root)) {
    throw new Error('@crestproject/hardhat: resolved path must be inside of project directory');
  }

  if (dir === root) {
    throw new Error('@crestproject/hardhat: resolved path must not be root directory');
  }

  return dir;
}

export async function clearDirectory(dir: string) {
  if (await fs.pathExists(dir)) {
    await fs.remove(dir);
  }
}

export async function createDirectory(dir: string) {
  if (!(await fs.pathExists(dir))) {
    await fs.mkdirp(dir);
  }
}

export async function prepareDirectory(dir: string, clear: boolean = false) {
  if (clear) {
    await clearDirectory(dir);
  }

  return createDirectory(dir);
}
