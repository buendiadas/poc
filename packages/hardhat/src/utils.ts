import fs from 'fs-extra';
import path from 'path';
import glob from 'glob';
import { HardhatProvider } from './provider';
import { validateConfig, resolveConfig, buildInfoDirName } from './imports';

export async function addCompilationResult(
  source: string,
  provider: HardhatProvider
) {
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const loaded = require(source);
  validateConfig(loaded);

  const config = resolveConfig(source, loaded);
  const dir = path.join(config.paths.artifacts, buildInfoDirName);
  const files = glob.sync('*.json', { cwd: dir, absolute: true });

  await Promise.all(
    files.map(async (file) => {
      const { input, output, solcVersion } = await fs.readJSON(file, {
        encoding: 'utf8',
      });

      await provider.send('hardhat_addCompilationResult', [
        solcVersion,
        input,
        output,
      ]);
    })
  );
}
