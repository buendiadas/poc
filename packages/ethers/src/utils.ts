import { ethers } from 'ethers';

export function propertyOf<TOr = any>(
  property: string,
  candidates: object[] = [],
): TOr {
  const obj = candidates.find((obj) => obj.hasOwnProperty(property));
  return (obj as any)?.[property] ?? undefined;
}

export function ensureInterface(
  fragments: string | (ethers.utils.Fragment | string)[],
) {
  if (ethers.utils.Interface.isInterface(fragments)) {
    return fragments;
  }

  return new ethers.utils.Interface(fragments);
}
