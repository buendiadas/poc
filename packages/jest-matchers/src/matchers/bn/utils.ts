import { ethers } from 'ethers';

export function printBigNumber(value: ethers.BigNumberish) {
  const bn = ethers.BigNumber.from(value);
  return `${bn.toString()} (${bn.toHexString()})`;
}

export type MatcherCallback = (
  received: ethers.BigNumber,
  expected: ethers.BigNumber,
) => jest.CustomMatcherResult;

export function ensureBigNumbers(
  context: jest.MatcherContext,
  received: any,
  expected: any,
  callback: MatcherCallback,
) {
  let receivedBn: ethers.BigNumber;
  let expectedBn: ethers.BigNumber;

  try {
    receivedBn = ethers.BigNumber.from(received);
  } catch {
    return forceFail(context, received, 'received');
  }

  try {
    expectedBn = ethers.BigNumber.from(expected);
  } catch {
    return forceFail(context, expected, 'expected');
  }

  return callback(receivedBn, expectedBn);
}

function forceFail(context: jest.MatcherContext, value: any, name: string) {
  const pass = context.isNot ? true : false;
  const message = () =>
    `The ${name} value is not numberish:\n\n` +
    `  ${context.utils.printReceived(value)}`;

  return { pass, message };
}
