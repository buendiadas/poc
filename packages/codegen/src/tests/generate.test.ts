import path from 'path';
import { ethers } from 'ethers';
import { formatOutput } from '../utils';
import {
  generateContractForSolidityArtifact,
  generateContractForSignatures,
} from '../generate';

describe('code generator', () => {
  it('generates code for solidity artifacts', () => {
    const contract = require('./contracts/ERC20.json');
    const source = path.resolve(__dirname, 'contracts/ERC20.json');
    const destination = path.resolve(__dirname, 'ERC20.ts');

    let relative = path.relative(path.dirname(destination), source);
    if (!relative.startsWith('.')) {
      relative = `./${relative}`;
    }

    const abi = new ethers.utils.Interface(contract.abi);
    const output = generateContractForSolidityArtifact('ERC20', relative, abi);
    const formatted = formatOutput(output);

    expect(formatted).toMatchSnapshot();
  });

  it('generates code for ethers signatures', () => {
    const contract = require('./contracts/ERC20.json');
    const abi = new ethers.utils.Interface(contract.abi);
    const output = generateContractForSignatures('ERC20', abi);
    const formatted = formatOutput(output);

    expect(formatted).toMatchSnapshot();
  });
});
