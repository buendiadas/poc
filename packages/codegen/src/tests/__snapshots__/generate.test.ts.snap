// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`code generator generates code 1`] = `
"/* eslint-disable */
import { ethers } from 'ethers';
import {
  contract,
  ConcreteContract,
  Call,
  Send,
  Functions,
} from '@crestproject/ethers';

export type ERC20Constructor = (name: string, symbol: string) => void;
export type ERC20Type = ConcreteContract<ERC20Functions>;
export interface ERC20Functions extends Functions {
  allowance: Call<(owner: string, spender: string) => ethers.BigNumber>;
  'allowance(address,address)': Call<
    (owner: string, spender: string) => ethers.BigNumber
  >;
  approve: Send<(spender: string, amount: ethers.BigNumberish) => boolean>;
  'approve(address,uint256)': Send<
    (spender: string, amount: ethers.BigNumberish) => boolean
  >;
  balanceOf: Call<(account: string) => ethers.BigNumber>;
  'balanceOf(address)': Call<(account: string) => ethers.BigNumber>;
  decimals: Call<() => ethers.BigNumber>;
  'decimals()': Call<() => ethers.BigNumber>;
  decreaseAllowance: Send<
    (spender: string, subtractedValue: ethers.BigNumberish) => boolean
  >;
  'decreaseAllowance(address,uint256)': Send<
    (spender: string, subtractedValue: ethers.BigNumberish) => boolean
  >;
  increaseAllowance: Send<
    (spender: string, addedValue: ethers.BigNumberish) => boolean
  >;
  'increaseAllowance(address,uint256)': Send<
    (spender: string, addedValue: ethers.BigNumberish) => boolean
  >;
  name: Call<() => string>;
  'name()': Call<() => string>;
  symbol: Call<() => string>;
  'symbol()': Call<() => string>;
  totalSupply: Call<() => ethers.BigNumber>;
  'totalSupply()': Call<() => ethers.BigNumber>;
  transfer: Send<(recipient: string, amount: ethers.BigNumberish) => boolean>;
  'transfer(address,uint256)': Send<
    (recipient: string, amount: ethers.BigNumberish) => boolean
  >;
  transferFrom: Send<
    (sender: string, recipient: string, amount: ethers.BigNumberish) => boolean
  >;
  'transferFrom(address,address,uint256)': Send<
    (sender: string, recipient: string, amount: ethers.BigNumberish) => boolean
  >;
}

export const ERC20 = contract.fromSolidity<ERC20Functions, ERC20Constructor>(
  require('./contracts/ERC20.json'),
);
"
`;
